NUI-611 remove manual set up of yaml file for the devtool

DrGirlfriend opened the pull request:
## Description

**This is backwards compatible because only functionality was introduced - no changes to any other repo or package library**

* Accepts integration file in JSON (downloaded from i/o console) or YAML (based on Asset Compute format) formats
* If JSON, a private key file is required
* If Integration file is not directly passed, ENV will be used `ASSET_COMPUTE_INTEGRATION_FILE_PATH'
* If JSON and private key is not passed directly, ENV will be used `ASSET_COMPUTE_PRIVATE_KEY_FILE_PATH`

100% test coverage for `lib/IntegrationConfiguration.js`

[NUI-611 remove manual set up of yaml file for the devtool](https://jira.corp.adobe.com/browse/NUI-611)

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

tmathern commented on lib/IntegrationConfiguration.js:
```diff
@@ -0,0 +1,104 @@
+    privateKeyPath = privateKeyFile || process.env.ASSET_COMPUTE_PRIVATE_KEY_FILE_PATH;
+    integrationPath = integrationFile || process.env.ASSET_COMPUTE_INTEGRATION_FILE_PATH;
+
+    return validate();
```
You could make the validate function async...

tmathern commented on lib/IntegrationConfiguration.js:
```diff
@@ -0,0 +1,104 @@
+
+function validate() {
+
+    if (!fs.existsSync(integrationPath)) {
```
by replacing `existsSync` with another call (`stat` or `lstat` if you stay with only node's `fs`. There are more options if you decide to use `fs-extra`, like `pathExists`).

tmathern commented on lib/IntegrationConfiguration.js:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    json.privateKey = validatePrivateKeyFile();
+    return convert(json);
```
I like conversion function to indicate what they convert `from` and `to`. Can you change the name so it is obvious?

DrGirlfriend replied:
I need ideas O_O the long name would be
`convertIOConsoleJsonToAssetComputeYamlConfiguration()`

I think `convertJsonToYaml()` is a bit confusing because it's not referring to what json it's converting from and we aren't really making a yaml file, just returning a differently structured object that resembles asset compute's yaml integration configuration file as an object

tmathern replied:
`convertConfigJsonToYaml` would be good enough for me!

tmathern commented on package.json:
```diff
@@ -12,6 +12,9 @@
         "@adobe/asset-compute-events-client": "^1.0.1",
         "@adobe/jwt-auth": "^0.3.1",
         "@adobe/node-fetch-retry": "^1.0.1",
+        "fs-extra": "^9.0.0",
```
nitpick: 9.0.1 as of today.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+
+        try {
+            await IntegrationConfiguration.getConfiguration(jsonFile, "fake-file.file");
+        } catch (e) {
```
We are looking for an exception to be thrown. It would be better to use assert.throws (https://nodejs.org/api/assert.html#assert_assert_throws_fn_error_message). 

Some errors have timestamps. You can control the timestamp value by using sinon fake timers: https://sinonjs.org/releases/v1.17.6/fake-timers/. This makes the content of the expected error fully known.

DrGirlfriend replied:
Would you accept `assert.rejects()` instead?

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+            actual = e;
+        }
+
+        assert.strictEqual(actual.name, expected.name);
```
Also, you do not reset `actual` before running this. So you may have previous content in there that could lead to inconclusive test results.
(Also, rather use assert.throws).

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+            actual = e;
+        }
+
+        assert.strictEqual(actual.name, expected.name);
```
Dito.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+        assert.strictEqual(actual.message, expected.message);
+    });
+
+    it('Input file is neither JSON nor YAML', async () => {
```
- Use assert.throws to check the proper exception is thrown.
- Be careful about state, make sure that the variables you check are set up properly, especially if you run multiple attempts in the same test. Here, `actual` may have old content that render the test inconclusive because `actual` is not resetted. 
- Use sinon fake timers (sinon.clock) to control time values if you need to.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+
+
+        process.env.ASSET_COMPUTE_INTEGRATION_FILE_PATH = yamlFile;
+        actual = await IntegrationConfiguration.getConfiguration();
```
Put this in another test (a separate `it(...)`. This will ensure there are no leftovers in `actual` from the test above. Leftovers render the test inconclusive: the test may pass because of old `actual` values.

The test title could be `it('Input file is YAML from env variable'`.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+        assert.strictEqual(actual.technicalAccount.privateKey, expected.technicalAccount.privateKey);
+
+
+        process.env.ASSET_COMPUTE_INTEGRATION_FILE_PATH = jsonFile;
```
Put this in another test (a separate it(...). This will ensure there are no leftovers in actual from the test above. Leftovers render the test inconclusive: the test may pass because of old actual values (for instance, if loading from the env variable set values fails silently).

The test title could be `it('Input file is JSON from env variable with private key')`.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+        assert.strictEqual(actual.technicalAccount.privateKey, expected.technicalAccount.privateKey);
+
+
+        process.env.ASSET_COMPUTE_PRIVATE_KEY_FILE_PATH = privateKeyFile;
```
Put this in another test (a separate it(...). This will ensure there are no leftovers in actual from the test above. Leftovers render the test inconclusive: the test may pass because of old actual values (for instance, if loading from the env variable set values fails silently).

The test title could be `it('Input file is JSON from env variable')`.

tmathern commented on test/integrationConfiguration.test.js:
```diff
@@ -0,0 +1,217 @@
+            actual = e;
+        }
+
+        assert.strictEqual(actual.name, expected.name);
```
Put this in its own test section (own `it(...)) to ensure there are no leftovers in `actual` from the test L192 to 203.

jdelbick commented:
This looks great! Definitely will be a big improvement on the developer experience. 

There's one more option we discovered yesterday that simplifies this even more that we should support:

If a developer logs in during `aio app init`, the credentials (besides the private key for now), will be directly in the `.env` file. It would be nice for the client or the devtool (not sure which should take care of this), to read these credentials directly from the `.env` file (no need for downloading credentials JSON). The `.env` file will look something like this:
```
AIO_$ims_Project__xxxxxx_client__id=xxxxxxxxx
AIO_$ims_Project__xxxxxx_client__secret=xxxxxxxe
AIO_$ims_Project__xxxxxx_technical__account__email=xxxxxx@techacct.adobe.com
AIO_$ims_Project__xxxxxx_technical__account__id=xxxxxxx@techacct.adobe.com
AIO_$ims_Project__xxxxxx_meta__scopes=["asset_compute_meta","ent_adobeio_sdk","event_receiver_api"]
AIO_$ims_Project__xxxxxx_ims__org__id=xxxxxxxx@AdobeOrg
```
then the user would only have to set the `ASSET_COMPUTE_PRIVATE_KEY_PATH` manually.

jdelbick commented:
@DrGirlfriend I think the unit tests are failing on Linux due to the name of the file being off:
https://travis-ci.com/github/adobe/asset-compute-client/jobs/353633281#L230
it is looking for this file: `../lib/integrationConfiguration` but the name of the file checked in is `../lib/IntegrationConfiguration`

tmathern commented:
> This looks great! Definitely will be a big improvement on the developer experience.
> 
> There's one more option we discovered yesterday that simplifies this even more that we should support:
> 
> If a developer logs in during `aio app init`, the credentials (besides the private key for now), will be directly in the `.env` file. It would be nice for the client or the devtool (not sure which should take care of this), to read these credentials directly from the `.env` file (no need for downloading credentials JSON). The `.env` file will look something like this:
> 
> ```
> AIO_$ims_Project__xxxxxx_client__id=xxxxxxxxx
> AIO_$ims_Project__xxxxxx_client__secret=xxxxxxxe
> AIO_$ims_Project__xxxxxx_technical__account__email=xxxxxx@techacct.adobe.com
> AIO_$ims_Project__xxxxxx_technical__account__id=xxxxxxx@techacct.adobe.com
> AIO_$ims_Project__xxxxxx_meta__scopes=["asset_compute_meta","ent_adobeio_sdk","event_receiver_api"]
> AIO_$ims_Project__xxxxxx_ims__org__id=xxxxxxxx@AdobeOrg
> ```
> 
> then the user would only have to set the `ASSET_COMPUTE_PRIVATE_KEY_PATH` manually.
> 
> this can be tackled in a different PR btw^^

There is already a bug-spike related to that point: [NUI-656](https://jira.corp.adobe.com/browse/NUI-656)

DrGirlfriend commented:
> @DrGirlfriend I think the unit tests are failing on Linux due to the name of the file being off:
> https://travis-ci.com/github/adobe/asset-compute-client/jobs/353633281#L230
> it is looking for this file: `../lib/integrationConfiguration` but the name of the file checked in is `../lib/IntegrationConfiguration`

Which is really strange because on my system I have `lib/integrationConfiguration`
<img width="337" alt="Screen Shot 2020-06-25 at 1 56 16 PM" src="https://user-images.githubusercontent.com/1933320/85789427-cc565080-b6eb-11ea-8c30-f26a1faf1ea6.png">

And `git status` says something else O_O
```
dhendric@Deirdras-MBP:~/working/ADOBE/asset-compute-client$ git status
On branch yamlconverter
Your branch is up to date with 'origin/yamlconverter'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   lib/IntegrationConfiguration.js
        modified:   package.json
        modified:   test/integrationConfiguration.test.js

no changes added to commit (use "git add" and/or "git commit -a")
```

Let me see if I can fix that via terminal `mv`

DrGirlfriend commented:
Figured it out and fixed it - GIT didn't acknowledge the rename so I `git rm` and then `git add` the file

jdelbick commented on README.md:
```diff
@@ -75,7 +77,9 @@ This function creates a new instance of `AssetComputeClient` and calls the `.reg
-    const integration = yaml.safeLoad(await fs.readFile("integration.yaml", "utf-8"));
+    //If integration file is json, a private key file must also be provided
+    const integrationFilePath = "/path/to/integration/file"; // Either json or yaml format
+    const integration = AssetComputeClient.getConfiguration(integrationFilePath[, privateKeyFile]);
```
maybe I am missing something, but where is this function inside `AssetComputeClient`?

jdelbick replied:
also after this is added to the `AssetComputeClient` class, please add a unit test for it :)

DrGirlfriend replied:
Changed it around

jdelbick commented on index.js:
```diff
@@ -14,11 +14,12 @@
 const { AssetComputeEventEmitter } = require("./lib/eventemitter");
-const { createAssetComputeClient, AssetComputeClient } = require("./lib/client");
+const { AssetComputeClient } = require("./lib/client");
+const { YamlConverter } = require('./lib/yamlconverter');
```
I think `IntegrationConfiguration` replaces this. It wouldn't be a bad idea to export the integration config anyways

tmathern commented:
LGTM regarding the changes I asked for.

jdelbick commented on lib/integrationConfiguration.js:
```diff
@@ -0,0 +1,100 @@
+    return fs.readFile(privateKeyPath, 'utf8');
+}
+
+function convertConfigJsonToYaml(json) {
```
btw, I would rename this function. The name seems to imply we are converting a JSON object into a YAML format. Instead, we are adjusting the format of the JSON object...maybe something like `convertConfigFormat`?

DrGirlfriend replied:
@tmathern ^^

tmathern replied:
`convertConfigFormat` was eliminated previously. We take a json, and convert it to a (previously) yaml format. If you want to rename it again, use `mapConfiguration`. So nobody complains about a possible notion of conversion. Does this sound like a deal?

jdelbick replied:
Sure "map" is more accurate.

jdelbick commented on package.json:
```diff
@@ -12,19 +12,22 @@
         "@adobe/node-fetch-retry": "^1.0.1",
+        "fs-extra": "^9.0.1",
+        "js-yaml": "^3.14.0",
+        "json-to-pretty-yaml": "^1.2.2",
```
looks like `json-to-pretty-yaml` is not used anywhere. I think we can remove this

jdelbick commented on test/resources/privatekey.pem:
```diff
@@ -0,0 +1 @@
+PRIVATE KEY FILE
```
looks like for some reason windows adds an extra `\r` somewhere in the private key file and that's why the tests are failing on windows: https://travis-ci.com/github/adobe/asset-compute-client/jobs/353755542#L139

DrGirlfriend replied:
blarg

tmathern replied:
(Trivia of the day: Windows created and Linux created files have different line endings. `\r` corresponds to CR, and CR LF is the new line on Windows machines (aka \r\n). Historically, this comes from commands send to printers when printers/tty started to exist.
That is why there is a setting somewhere in git to choose which line ending to use, for instance.)
